### ISSUE TITLE
qbec is trying to remove read-only fields from Service and PVC

### ISSUE THREAD

**Issue by Andor** [splunk/qbec#33]
These object were created with `qbec show -S infra | kubectl apply -f -`
```
andor@titude:~/work/kube$ qbec diff infra -c grafana
setting cluster to gke__europe-north1_infra
setting context to gke__europe-north1_infra
cluster metadata load took 475ms
--- live persistentvolumeclaims grafana-storage -n default (source grafana) (source: fallback - live object with some attributes rem
oved)
+++ config persistentvolumeclaims grafana-storage -n default (source grafana)
@@ -2,14 +2,7 @@
 kind: PersistentVolumeClaim
 metadata:
   annotations:
-    kubectl.kubernetes.io/last-applied-configuration: |
-      {"apiVersion":"v1","kind":"PersistentVolumeClaim","metadata":{"annotations":{"qbec.io/component":"grafana"},"labels":{"qbec.io/application":"infra","qbec.io/environment":"infra"},"name":"grafana-storage","namespace":"default"},"spec":{"accessModes":["ReadWriteOnce"],"resources":{"requests":{"storage":"2Gi"}}}}
-    pv.kubernetes.io/bind-completed: "yes"
-    pv.kubernetes.io/bound-by-controller: "yes"
     qbec.io/component: grafana
-    volume.beta.kubernetes.io/storage-provisioner: kubernetes.io/gce-pd
-  finalizers:
-  - kubernetes.io/pvc-protection
   labels:
     qbec.io/application: infra
     qbec.io/environment: infra
@@ -18,11 +11,7 @@
 spec:
   accessModes:
   - ReadWriteOnce
-  dataSource: null
   resources:
     requests:
       storage: 2Gi
-  storageClassName: standard
-  volumeMode: Filesystem
-  volumeName: pvc-a84ff09e-9266-11e9-ab85-42010aa60144
 


--- live services grafana -n default (source grafana) (source: fallback - live object with some attributes removed)
+++ config services grafana -n default (source grafana)
@@ -2,8 +2,6 @@
 kind: Service
 metadata:
   annotations:
-    kubectl.kubernetes.io/last-applied-configuration: |
-      {"apiVersion":"v1","kind":"Service","metadata":{"annotations":{"qbec.io/component":"grafana"},"labels":{"app":"grafana","qbec.io/application":"infra","qbec.io/environment":"infra"},"name":"grafana","namespace":"default"},"spec":{"ports":[{"name":"http","nodePort":30910,"port":3000,"targetPort":"http"}],"selector":{"app":"grafana"},"type":"NodePort"}}
     qbec.io/component: grafana
   labels:
     app: grafana
@@ -12,16 +10,12 @@
   name: grafana
   namespace: default
 spec:
-  clusterIP: 10.3.241.2
-  externalTrafficPolicy: Cluster
   ports:
   - name: http
     nodePort: 30910
     port: 3000
-    protocol: TCP
     targetPort: http
   selector:
     app: grafana
-  sessionAffinity: None
   type: NodePort
 


waiting for deletion list to be returned
server objects load took 1.915s
---
stats:
  changes:
  - persistentvolumeclaims grafana-storage -n default (source grafana)
  - services grafana -n default (source grafana)
  same: 4

✘ 2 object(s) different
command took 7.31s
```
And when I'm trying to apply the changes:
```
andor@titude:~/work/kube$ qbec apply --yes infra -c grafana 
setting cluster to gke__europe-north1_infra
setting context to gke__europe-north1_infra
cluster metadata load took 282ms

will synchronize 6 object(s)

✘ sync persistentvolumeclaims grafana-storage -n default (source grafana): PersistentVolumeClaim "grafana-storage" is invalid: spec: Forbidden: is immutable after creation except resources.requests for bound claims
command took 6.51s
```

Versions:
```
andor@titude:~/work/kube$ qbec version
qbec version: 0.7.0
jsonnet version: v0.11.2
go version: 1.12.5
commit: a91b4a2
andor@titude:~/work/kube$ kubectl version
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.7", GitCommit:"4683545293d792934a7a7e12f2cc47d20b2dd01b", GitTreeState:"clean", BuildDate:"2019-06-06T01:46:52Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"13+", GitVersion:"v1.13.6-gke.13", GitCommit:"fcbc1d20b6bca1936c0317743055ac75aef608ce", GitTreeState:"clean", BuildDate:"2019-06-19T20:50:07Z", GoVersion:"go1.11.5b4", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by kvaps** [comment]
Qbec should probably fallback to `kubectl.kubernetes.io/last-applied-configuration` and use it as base for diff, instead using whole manifest to avoid situations like this.

**Comment by gotwarlost** [comment]
could you paste the output of `apply --dry-run` for the object so I can tell what it is trying to patch. Unfortunately qbec only prints the patch when it succeeds which is itself a separate bug.

**Comment by Andor** [comment]
No idea how to explicitly disable colors, so here it is:
```[1msetting cluster to gke__europe-north1_infra
[0m[1msetting context to gke__europe-north1_infra
[0m[2mcluster metadata load took 289ms
[0m[1m[dry-run] sync persistentvolumeclaims grafana-storage -n default (source grafana)
[0mkind: application/strategic-merge-patch+json
operation: update object
patch: |-
  {
      "metadata": {
          "annotations": {
              "kubectl.kubernetes.io/last-applied-configuration": null,
              "pv.kubernetes.io/bind-completed": null,
              "pv.kubernetes.io/bound-by-controller": null,
              "qbec.io/last-applied": "H4sIAAAAAAAA/1SOsU4sMQxF+/cZt84DQZmWggqBKJYCUXgznlVEYmdjzzar/DuaGQGis32uju8V1PKBu2UVRFzuEPCZZULEy3o1Z/GDlqXyQ6FcEVDZaSInxCtIRJ08q9i6no+cbrLeJq1NhcURceo0kxBGQKEjlz9Baq3ktAkQkWXuhPBDWS65q9RdtNMRIFT5V/zfXDudGDuwRmmlE8+0FF/z1jhtZVNisyed2BDf8co0vfXs/CyJ8RHQ2XTpibeGnc8Lm2/z94eI+8eMMcb49wUAAP//AQAA//+osorhOAEAAA==",
              "volume.beta.kubernetes.io/storage-provisioner": null
          },
          "finalizers": null
      },
      "spec": {
          "dataSource": null,
          "storageClassName": null,
          "volumeMode": null,
          "volumeName": null
      }
  }
source: open API

[1m[dry-run] sync deployments grafana -n default (source grafana)
[0mkind: application/strategic-merge-patch+json
operation: update object
patch: |-
  {
      "spec": {
          "template": {
              "spec": {
                  "$setElementOrder/containers": [
                      {
                          "name": "grafana"
                      }
                  ],
                  "containers": [
                      {
                          "$setElementOrder/volumeMounts": [
                              {
                                  "mountPath": "/var/lib/grafana"
                              },
                              {
                                  "mountPath": "/etc/grafana/provisioning/datasources"
                              },
                              {
                                  "mountPath": "/etc/grafana/provisioning/dashboards"
                              }
                          ],
                          "name": "grafana",
                          "volumeMounts": [
                              {
                                  "mountPath": "/var/lib/grafana",
                                  "readOnly": false
                              },
                              {
                                  "mountPath": "/etc/grafana/provisioning/datasources",
                                  "readOnly": false
                              },
                              {
                                  "mountPath": "/etc/grafana/provisioning/dashboards",
                                  "readOnly": false
                              }
                          ]
                      }
                  ]
              }
          }
      }
  }
source: open API

[1m[dry-run] sync services grafana -n default (source grafana)
[0mkind: application/strategic-merge-patch+json
operation: update object
patch: |-
  {
      "metadata": {
          "annotations": {
              "kubectl.kubernetes.io/last-applied-configuration": null,
              "qbec.io/last-applied": "H4sIAAAAAAAA/2SPwUoEMRBE735GnYPO4sl8hAiCF/HQm+ldgzPdbdIOyJB/l8ywsuKxq+q9kBVk+YVLzSqIWA4I+MgyIuKZy5ITI2Bmp5GcEFeQiDp5Vqn9/Dxyus16l3Q2FRZHxLnQiYTQAiY68rQNyeyqCr8gmU05bUJEZDmV65ZlyUVl3sV72wKEZv5j60E1Sj0d+URfk/ddNU79cdPiFfF1vZDv7tYxHflJiyPeDw+HIWzDfgxDgFM5s+/1DrS3gMoTJ9fy/08twL+t2x8v2tZufgAAAP//AQAA//8MWQ4HYQEAAA=="
          }
      },
      "spec": {
          "$setElementOrder/ports": [
              {
                  "port": 3000
              }
          ],
          "clusterIP": null,
          "externalTrafficPolicy": null,
          "ports": [
              {
                  "port": 3000,
                  "protocol": null
              }
          ],
          "sessionAffinity": null
      }
  }
source: open API

[2mwaiting for deletion list to be returned
[0m[2mserver objects load took 1.915s
[0m---
stats:
  same: 3
  updated:
  - persistentvolumeclaims grafana-storage -n default (source grafana)
  - deployments grafana -n default (source grafana)
  - services grafana -n default (source grafana)

[1m** dry-run mode, nothing was actually changed **
[0m[2mcommand took 5.5s
[0m```

**Comment by gotwarlost** [comment]
sorry for the delay in response. Could you show me the output of the PVC object using `qbec show` so I can see what the original looks like and repro this bug. Thanks!

**Comment by Andor** [comment]
@gotwarlost 
```
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  annotations:
    qbec.io/component: grafana
  labels:
    qbec.io/application: infra
    qbec.io/environment: infra
  name: grafana-storage
  namespace: default
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 2Gi
```

**Comment by gotwarlost** [comment]
Thanks for all the info. I can verify that you can create this object with `kubectl create` and `kubectl apply` will work fine but `qbec apply` won't. I need to look at the latest patch algorithm implemented by kubectl and upgrade qbec with it

**Comment by gotwarlost** [comment]
After chasing all forms of subtle differences between how kubectl and qbec patches work, the bug turned out to be something really stupid (which must have occurred when I took internal code, "simplified" it and open sourced).

https://github.com/splunk/qbec/blob/4a82732cfcb8f827220b2f920dba6a2695c29aef/internal/remote/pristine.go#L137-L140

Anyway, fix is on its way: https://github.com/splunk/qbec/pull/52

only to be committed after unit tests are added for that file

**Comment by gotwarlost** [comment]
this fix will also have an enhancement suggested by @kvaps, which is to use the kubectl annotation, if present, to compute diffs and patches.

**Comment by gotwarlost** [comment]
available in v0.7.4 https://github.com/splunk/qbec/releases/tag/v0.7.4
